codegen/src/SchemaProvider.scala

```diff
@@ -2,89 +2,144 @@ package besom.codegen
-      os.makeDir.all(schemaFilePath / os.up)
-      os.proc("pulumi", "package", "get-schema", schemaSource).call(stdout = schemaFilePath)
+      logger.debug(
+        s"Downloading schema for '${metadata.name}:${metadata.version}' into '${schemaFilePath.relativeTo(os.pwd)}'"
+      )
+
+      val schemaSource =
+        if (metadata.version.isDefault)
+          metadata.name
+        else
+          s"${metadata.name}@${metadata.version}"
+
+      val installCmd: List[String] =
+        List("pulumi", "plugin", "install", "resource", metadata.name, metadata.version) ++
+          metadata.server.map(url => List("--server", url)).getOrElse(List.empty)
+      os.proc(installCmd).call()
+
+      val schema =
+        try {
+          os.proc("pulumi", "package", "get-schema", schemaSource).call().out.text
+        } catch {
+          case e: os.SubprocessException =>
+            val msg =
+              s"Failed to download schema for ${metadata.name}:${metadata.version}' into '${schemaFilePath.relativeTo(os.pwd)}'"
+            logger.error(msg)
+            throw GeneralCodegenException(msg, e)
+        }
+
+      // parse and save the schema using path corrected for the actual name and version for the package
+      val pulumiPackage         = PulumiPackage.fromString(schema)
+      val reconciled            = pulumiPackage.toPackageMetadata(metadata.version)
+      val correctSchemaFilePath = schemaCacheDirPath / reconciled.name / reconciled.version / schemaFileName
+      os.write.over(correctSchemaFilePath, schema, createFolders = true)
+      (pulumiPackage, reconciled)
     } else {
-      logger.debug(s"Using cached schema for $providerName:$schemaVersion' from '${schemaFilePath.relativeTo(os.pwd)}'")
+      logger.debug(
+        s"Using cached schema for ${metadata.name}:${metadata.version}' from '${schemaFilePath.relativeTo(os.pwd)}'"
+      )
+      (PulumiPackage.fromFile(schemaFilePath), metadata)
     }
-
-    schemaFilePath
   }
 
-  def addSchemaFile(providerName: ProviderName, schemaVersion: SchemaVersion, content: os.Path): os.Path = {
-    val schemaFilePath = schemaCacheDirPath / providerName / schemaVersion / "schema.json"
-    os.copy.over(content, schemaFilePath, replaceExisting = true, createFolders = true)
-    schemaFilePath
+  private def pulumiPackage(schema: SchemaFile): (PulumiPackage, PackageMetadata) = {
+    // try to get fallback version information from the schema file path
+    val pathVersion = {
+      val directory = (schema / os.up).last
+      if (PackageVersion.isValid(directory)) PackageVersion(directory) else PackageVersion()
+    }
+
+    // parse and save the schema using path corrected for the actual name and version for the package
+    val pulumiPackage         = PulumiPackage.fromFile(schema)
+    val metadata              = pulumiPackage.toPackageMetadata(pathVersion)
+    val correctSchemaFilePath = schemaCacheDirPath / metadata.name / metadata.version / schemaFileName
```

Comment:
In what sense is this path `correct`? Is there any other path that is incorrect?

Reply:
there can be (and is often in tests) a defference of name and version between the arguments that are provided and the contents of the actual schema file e.g. multpile test schemas from upstream have name 'example' and no version